#### Quarterly Sales Tax Revenue Analysis

### [Data Source](https://tax.utah.gov/econstats/sales)

### Q4: October, November, December

### Notes:
- Q4 2020 excel sheet does not contain totals for the 'UNKNOWN/NONCLASSIFIABLE' category; respective column for 2019 was removed

In [1]:
import pandas as pd
import os
import numpy as np
import arcpy
from arcgis.features import GeoAccessor, GeoSeriesAccessor
arcpy.env.overwriteOutput = True

In [2]:
# show all columns
pd.options.display.max_columns = None

In [3]:
# Removes misc characters from string
def replace(string):
    string = string.replace("&", "")
    string = string.replace("-", " ")
    string = string.replace("(", "")
    string = string.replace(")", "")
    string = string.replace("/", " ")
    string = string.replace(",", "")
    string = string.replace(".", "")
    string = string.replace(" ", "_")
    return string


### Load Q4
2020 sales tax revenue data

In [4]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2020-04-quarterly-sales.xlsx'
q4_2020 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
q4_2020.drop(q4_2020.tail(2).index,inplace=True)

# forward fill values from merged cells
q4_2020['County'] = q4_2020['County'].fillna(method='ffill')
q4_2020['Location Code'] = q4_2020['Location Code'].fillna(method='ffill')
q4_2020['City'] = q4_2020['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
q4_2020.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
q4_2020['Rounded Up'] = q4_2020['Rounded Up'].map(crosswalk)

q4_2020.head()


County  Location Code         City  \
0  Beaver County         1002.0  Beaver City   
1  Beaver County         1002.0  Beaver City   
2  Beaver County         1002.0  Beaver City   
3  Beaver County         1002.0  Beaver City   
4  Beaver County         1002.0  Beaver City   

                    Economic Sector (NAICS Code)     2020Q4  Rounded Up  
0  AGRICULTURE, FORESTRY, FISHING & HUNTING (11)    30000.0        True  
1                                 UTILITIES (22)   700000.0        True  
2                              CONSTRUCTION (23)    93769.0       False  
3                          MANUFACTURING (31-33)  1909328.0       False  
4            WHOLESALE TRADE-DURABLE GOODS (423)   317274.0       False

In [5]:
# pivot the table to get sectors as columns with tax values by city
sectors_q4_2020 = pd.pivot_table(q4_2020,values = '2020Q4',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_q4_2020 = sectors_q4_2020.fillna(0)

In [6]:
sectors_q4_2020.columns

Index(['ACCOMMODATION (721)',
       'ADMIN. & SUPPORT & WASTE MANAG. & REMED. SERVICES (56)',
       'AGRICULTURE, FORESTRY, FISHING & HUNTING (11)',
       'ARTS, ENTERTAINMENT AND RECREATION (71)', 'CONSTRUCTION (23)',
       'EDUCATIONAL SERVICES (61)', 'FINANCE & INSURANCE (52)',
       'FOOD SERVICES & DRINKING PLACES (722)',
       'HEALTH CARE & SOCIAL ASSISTANCE (62)', 'INFORMATION (51)',
       'MANAGEMENT OF COMPANIES & ENTERPRISES (55)', 'MANUFACTURING (31-33)',
       'MINING, QUARRYING, & OIL & GAS EXTRACTION (21)',
       'OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)',
       'PRIOR-PERIOD PAYMENTS & REFUNDS', 'PRIVATE MOTOR VEHICLE SALES',
       'PROFESSIONAL, SCIENTIFIC & TECHNICAL SERVICES (54)',
       'PUBLIC ADMINISTRATION (92)', 'REAL ESTATE, RENTAL & LEASING (53)',
       'RETAIL-BUILD. MATERIAL, GARDEN EQUIP. & SUPPLIES DEALERS (444)',
       'RETAIL-CLOTHING & CLOTHING ACCESSORIES STORES (448)',
       'RETAIL-ELECTRONICS & APPLIANCE STORES (443)',
       

In [7]:
# fix column names
#new_names = [replace(item) for item in list(sectors_q4_2020.columns)]
new_names = ['L_ACCOMMODATION_20', 'S_ADMIN_SUPPORT_20','A_AG_WILDLIFE_20','L_CULTURAL_REC_20','C_CONSTRUCTION_20',
'E_EDUCATION_20','S_FINANCIAL_20','L_RSTRNT_BAR_20','H_HEALTH_CARE_20','S_IT_DATA_20','S_CORPORATE_MGMT_20',
'M_MANUFACTURING_20','E_EXTRACTION_20','O_OTHER_20','X_ADJUSTMENTS_20','R_AUTO_PRIVATE_20','S_PROF_TECH_SRV_20',
'G_GOVERNMENT_20','S_RENTAL_LEASING_20','R_BUILDING_SUPPLY_20','R_CLOTHING_20','R_ELECTRONICS_20','R_GROCERY_BEV_20',
'R_FURNITURE_20','R_GAS_STATIONS_20','R_GENERAL_RETAIL_20','R_HEALTH_RETAIL_20','R_OTHER_RETAIL_20','R_AUTO_RETAIL_20',
'R_NONSTORE_RETAIL_20','R_SPORT_HOBBY_20','L_SPECIAL_EVENT_20','W_DISTRIBUTION_20','X_UNKNOWN_20','U_UTILITIES_20',
'W_WHLSALE_DURABLE_20','S_WHLSLE_ETRADE_20','W_WHLSLE_NDURABLE_20']
sectors_q4_2020.columns = new_names

# get the total sales tax revenue
sectors_q4_2020['TOTAL_20'] = sectors_q4_2020[new_names].sum(axis=1)

sectors_q4_2020.head()

L_ACCOMMODATION_20  S_ADMIN_SUPPORT_20  A_AG_WILDLIFE_20  \
City                                                                      
American Fork            445919.0            946108.0          400000.0   
Beaver City             1263540.0             21241.0           30000.0   
Blanding                 674993.0             24381.0            1000.0   
Bountiful                200000.0           1670480.0            7550.0   
Brigham                  905882.0            196396.0            2000.0   

               L_CULTURAL_REC_20  C_CONSTRUCTION_20  E_EDUCATION_20  \
City                                                                  
American Fork          1383561.0         11147175.0         98622.0   
Beaver City               8000.0            93769.0         20000.0   
Blanding                 45000.0           212049.0         70000.0   
Bountiful              1218068.0          1718083.0         89674.0   
Brigham                 311891.0           921313.0         68816.0   

               S_FINANCIAL_20  L_RSTRNT_BAR_20  H_HEALTH_CARE_20  \
City                                                               
American Fork       3568796.0       27448933.0          649328.0   
Beaver City          515183.0        2642442.0           60000.0   
Blanding              17960.0         400000.0           40000.0   
Bountiful           1139404.0       13172819.0         1623303.0   
Brigham              232119.0        7240559.0          433033.0   

               S_IT_DATA_20  S_CORPORATE_MGMT_20  M_MANUFACTURING_20  \
City                                                                   
American Fork     9578219.0              25000.0           3582969.0   
Beaver City       1475591.0               1000.0           1909328.0   
Blanding           434186.0               1000.0            409957.0   
Bountiful         6279817.0              20000.0           3261099.0   
Brigham           3040482.0               4000.0           8030351.0   

               E_EXTRACTION_20  O_OTHER_20  X_ADJUSTMENTS_20  \
City                                                           
American Fork         150000.0  12001800.0         4157186.0   
Beaver City                0.0    928671.0          215431.0   
Blanding                   0.0    456694.0           -4831.0   
Bountiful              40000.0   6319648.0         1789498.0   
Brigham                 8000.0   3697620.0          537328.0   

               R_AUTO_PRIVATE_20  S_PROF_TECH_SRV_20  G_GOVERNMENT_20  \
City                                                                    
American Fork          6500000.0           7326059.0          20000.0   
Beaver City             400000.0            175210.0              0.0   
Blanding                700000.0            158873.0         500000.0   
Bountiful              4250000.0           4820458.0         300000.0   
Brigham                2000000.0            925025.0           1000.0   

               S_RENTAL_LEASING_20  R_BUILDING_SUPPLY_20  R_CLOTHING_20  \
City                                                                      
American Fork            3690672.0            23190264.0      9857552.0   
Beaver City               196120.0              873826.0       199170.0   
Blanding                   66774.0             2255336.0       121948.0   
Bountiful                2233518.0             2865086.0      4175511.0   
Brigham                  2226479.0             3503291.0       717404.0   

               R_ELECTRONICS_20  R_GROCERY_BEV_20  R_FURNITURE_20  \
City                                                                
American Fork         9775817.0         3874635.0       5670578.0   
Beaver City             41237.0         2000000.0         33318.0   
Blanding               108118.0         2000000.0         41293.0   
Bountiful             5053796.0        29675024.0       2085829.0   
Brigham               1773289.0        13453415.0        641342.0   

               R_GAS_STATIONS_20  R_GE

#### Load Q4 2019 sales tax revenue data

In [8]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2019-04-quarterly-sales.xlsx'
q4_2019 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
q4_2019.drop(q4_2019.tail(2).index,inplace=True)

# forward fill values from merged cells
q4_2019['County'] = q4_2019['County'].fillna(method='ffill')
q4_2019['Location Code'] = q4_2019['Location Code'].fillna(method='ffill')
q4_2019['City'] = q4_2019['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
q4_2019.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
q4_2019['Rounded Up'] = q4_2019['Rounded Up'].map(crosswalk)

q4_2019.tail()

County  Location Code         City  \
2410  Weber County        29040.0  South Ogden   
2411  Weber County        29040.0  South Ogden   
2412  Weber County        29040.0  South Ogden   
2413  Weber County        29040.0  South Ogden   
2414  Weber County        29040.0  South Ogden   

                          Economic Sector (NAICS Code)     2019Q4  Rounded Up  
2410  OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)  1391763.0       False  
2411                        PUBLIC ADMINISTRATION (92)     1000.0        True  
2412                       PRIVATE MOTOR VEHICLE SALES  1500000.0        True  
2413                               SPECIAL EVENT SALES    35000.0        True  
2414                   PRIOR-PERIOD PAYMENTS & REFUNDS   179101.0       False

In [13]:
# pivot the table to get sectors as columns with tax values by city
sectors_q4_2019 = pd.pivot_table(q4_2019,values = '2019Q4',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_q4_2019 = sectors_q4_2019.fillna(0)

# fix column names
#new_names = [replace(item) for item in list(sectors_q4_2019.columns)]
new_names = ['L_ACCOMMODATION_19', 'S_ADMIN_SUPPORT_19','A_AG_WILDLIFE_19','L_CULTURAL_REC_19','C_CONSTRUCTION_19',
'E_EDUCATION_19','S_FINANCIAL_19','L_RSTRNT_BAR_19','H_HEALTH_CARE_19','S_IT_DATA_19','S_CORPORATE_MGMT_19',
'M_MANUFACTURING_19','E_EXTRACTION_19','O_OTHER_19','X_ADJUSTMENTS_19','R_AUTO_PRIVATE_19','S_PROF_TECH_SRV_19',
'G_GOVERNMENT_19','S_RENTAL_LEASING_19','R_BUILDING_SUPPLY_19','R_CLOTHING_19','R_ELECTRONICS_19','R_GROCERY_BEV_19',
'R_FURNITURE_19','R_GAS_STATIONS_19','R_GENERAL_RETAIL_19','R_HEALTH_RETAIL_19','R_OTHER_RETAIL_19','R_AUTO_RETAIL_19',
'R_NONSTORE_RETAIL_19','R_SPORT_HOBBY_19','L_SPECIAL_EVENT_19','W_DISTRIBUTION_19','X_UNKNOWN_19','U_UTILITIES_19',
'W_WHLSALE_DURABLE_19','S_WHLSLE_ETRADE_19','W_WHLSLE_NDURABLE_19']
sectors_q4_2019.columns = new_names

# drop unknown column since its not present in 2020
#sectors_q4_2019.drop('X_UNKNOWN_19', axis=1, inplace=True)
#new_names.remove('X_UNKNOWN_19')

# get the total sales tax revenue
sectors_q4_2019['TOTAL_19'] = sectors_q4_2019[new_names].sum(axis=1)

sectors_q4_2019.head()

L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
City                                                                      
American Fork            554928.0            698435.0          200000.0   
Beaver City             1056763.0             20808.0           15000.0   
Blanding                 761898.0             61901.0            1000.0   
Bountiful                150000.0           2291194.0            2000.0   
Brigham                  984729.0            222487.0            1000.0   

               L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  \
City                                                                  
American Fork          1565976.0          9116585.0         58968.0   
Beaver City               9000.0            70000.0         25000.0   
Blanding                 70000.0           276229.0        150000.0   
Bountiful              1098539.0          1894207.0         91173.0   
Brigham                 260399.0           625648.0         67697.0   

               S_FINANCIAL_19  L_RSTRNT_BAR_19  H_HEALTH_CARE_19  \
City                                                               
American Fork       4184777.0       24894878.0          553191.0   
Beaver City          515603.0        2206640.0           45000.0   
Blanding              21221.0         450000.0           60000.0   
Bountiful           1332378.0       12786832.0         1039103.0   
Brigham              455220.0        6659356.0          308659.0   

               S_IT_DATA_19  S_CORPORATE_MGMT_19  M_MANUFACTURING_19  \
City                                                                   
American Fork    10533285.0               9000.0           2593500.0   
Beaver City        520147.0               3000.0           1497139.0   
Blanding           431669.0               4000.0             94027.0   
Bountiful         5303124.0               9000.0           2863668.0   
Brigham           2858043.0              25000.0           6604323.0   

               E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
City                                                           
American Fork          80000.0  10867304.0         3517109.0   
Beaver City                0.0    782177.0         -555466.0   
Blanding                   0.0    277858.0         1129905.0   
Bountiful              15000.0   6239951.0         2904446.0   
Brigham                 1000.0   3293364.0        -1541163.0   

               R_AUTO_PRIVATE_19  S_PROF_TECH_SRV_19  G_GOVERNMENT_19  \
City                                                                    
American Fork          2500000.0           4072879.0          15000.0   
Beaver City             300000.0            123929.0              0.0   
Blanding                600000.0            152346.0         600000.0   
Bountiful              3250000.0           4140342.0         250000.0   
Brigham                1500000.0            870400.0           1000.0   

               S_RENTAL_LEASING_19  R_BUILDING_SUPPLY_19  R_CLOTHING_19  \
City                                                                      
American Fork            3694025.0            18791011.0      9024897.0   
Beaver City               110090.0              566512.0       117279.0   
Blanding                   74215.0             1717197.0        83400.0   
Bountiful                2171159.0             1061688.0      4059340.0   
Brigham                  1395475.0             1714961.0       583848.0   

               R_ELECTRONICS_19  R_GROCERY_BEV_19  R_FURNITURE_19  \
City                                                                
American Fork        11596032.0         4074889.0       4627530.0   
Beaver City             30530.0         1750000.0         17014.0   
Blanding               143702.0         1750000.0          8692.0   
Bountiful             3841132.0        28438413.0       1633886.0   
Brigham               1520067.0        12372772.0        525222.0   

               R_GAS_STATIONS_19  R_GE

### Calculate the differences 
*2020 - 2019 = Diff*

In [14]:
# Check join output shape
print(sectors_q4_2019.shape)
print(sectors_q4_2020.shape)

(66, 39)
(66, 39)


In [15]:
# merge the 2019-2020 difference with the sdf
sectors_q4_diff = pd.DataFrame(sectors_q4_2020.values - sectors_q4_2019.values)

# add back the column names
new_names = ['L_ACCOMMODATION_D20', 'S_ADMIN_SUPPORT_D20','A_AG_WILDLIFE_D20','L_CULTURAL_REC_D20','C_CONSTRUCTION_D20',
'E_EDUCATION_D20','S_FINANCIAL_D20','L_RSTRNT_BAR_D20','H_HEALTH_CARE_D20','S_IT_DATA_D20','S_CORPORATE_MGMT_D20',
'M_MANUFACTURING_D20','E_EXTRACTION_D20','O_OTHER_D20','X_ADJUSTMENTS_D20','R_AUTO_PRIVATE_D20','S_PROF_TECH_SRV_D20',
'G_GOVERNMENT_D20','S_RENTAL_LEASING_D20','R_BUILDING_SUPPLY_D20','R_CLOTHING_D20','R_ELECTRONICS_D20','R_GROCERY_BEV_D20',
'R_FURNITURE_D20','R_GAS_STATIONS_D20','R_GENERAL_RETAIL_D20','R_HEALTH_RETAIL_D20','R_OTHER_RETAIL_D20','R_AUTO_RETAIL_D20',
'R_NONSTORE_RETAIL_D20','R_SPORT_HOBBY_D20','L_SPECIAL_EVENT_D20','W_DISTRIBUTION_D20','X_UNKNOWN_D20','U_UTILITIES_D20',
'W_WHLSALE_DURABLE_D20','S_WHLSLE_ETRADE_D20','W_WHLSLE_NDURABLE_D20', 'TOTAL_D20']

sectors_q4_diff.columns = new_names

# add back cities as index
sectors_q4_diff['City'] = sectors_q4_2020.index
sectors_q4_diff = sectors_q4_diff.set_index('City')

#check table
sectors_q4_diff.head(10)

L_ACCOMMODATION_D20  S_ADMIN_SUPPORT_D20  \
City                                                           
American Fork                 -109009.0             247673.0   
Beaver City                    206777.0                433.0   
Blanding                       -86905.0             -37520.0   
Bountiful                       50000.0            -620714.0   
Brigham                        -78847.0             -26091.0   
Cedar City                     -41175.0             671086.0   
Centerville                     25000.0              48956.0   
Clearfield                      50000.0            -233464.0   
Cottonwood Heights             129633.0             -20690.0   
Delta                           50000.0              -1410.0   

                    A_AG_WILDLIFE_D20  L_CULTURAL_REC_D20  C_CONSTRUCTION_D20  \
City                                                                            
American Fork                200000.0           -182415.0           2030590.0   
Beaver City                   15000.0             -1000.0             23769.0   
Blanding                          0.0            -25000.0            -64180.0   
Bountiful                      5550.0            119529.0           -176124.0   
Brigham                        1000.0             51492.0            295665.0   
Cedar City                    -4710.0            114423.0            385898.0   
Centerville                   -3000.0             55000.0           -236875.0   
Clearfield                     2000.0           -373752.0            534101.0   
Cottonwood Heights             2000.0           -742029.0           -733031.0   
Delta                             0.0            -10000.0          -3475527.0   

                    E_EDUCATION_D20  S_FINANCIAL_D20  L_RSTRNT_BAR_D20  \
City                                                                     
American Fork               39654.0        -615981.0         2554055.0   
Beaver City                 -5000.0           -420.0          435802.0   
Blanding                   -80000.0          -3261.0          -50000.0   
Bountiful                   -1499.0        -192974.0          385987.0   
Brigham                      1119.0        -223101.0          581203.0   
Cedar City                -148550.0          77353.0         2374878.0   
Centerville                 32029.0         -50352.0          413992.0   
Clearfield                   9142.0         -86233.0          391381.0   
Cottonwood Heights         -78076.0           7387.0         -684317.0   
Delta                        1000.0          19576.0           50151.0   

                    H_HEALTH_CARE_D20  S_IT_DATA_D20  S_CORPORATE_MGMT_D20  \
City                                                                         
American Fork                 96137.0      -955066.0               16000.0   
Beaver City                   15000.0       955444.0               -2000.0   
Blanding                     -20000.0         2517.0               -3000.0   
Bountiful                    584200.0       976693.0               11000.0   
Brigham                      124374.0       182439.0              -21000.0   
Cedar City                     5390.0       -74291.0               -1000.0   
Centerville                   46730.0     -1913699.0                   0.0   
Clearfield                    -6320.0      1116903.0              200000.0   
Cottonwood Heights           -50553.0       928925.0             -100000.0   
Delta                        -20000.0        75196.0               -7000.0   

                    M_MANUFACTURING_D20  E_EXTRACTION_D20  O_OTHER_D20  \
City                                                                     
American Fork                  989469.0           70000.0    1134496.0   
Beaver City                    412189.0               0.0     146494.0   
Blanding                       315930.0               0.0     178836.0   
Bountiful                      397431.0           25000.0      79697.0   
Brigham 

### Load municipalities and townships shapefile and join to formatted sales tax data

In [16]:
# load cities shapefile into pandas spatial dataframe
cities_shp = '.\\Inputs\\Cities.shp'
cities_sdf = pd.DataFrame.spatial.from_featureclass(cities_shp)

# Rename columns
cities_sdf.columns = ['FID', 'NAME', 'SOURCE', 'SALESTAXID', 'POPLASTCEN', 'POPLASTEST',
       'AREA_SQMI', 'SHAPE']

In [17]:
# merge the 2020 sdf with the pivoted data
cities_sdf2 = cities_sdf.merge(sectors_q4_2019, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_q4_2020, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_q4_diff, left_on= "NAME", right_on="City", how="inner")
cities_sdf2.tail(10)


FID              NAME  SOURCE SALESTAXID  POPLASTCEN  POPLASTEST  \
55  222          Syracuse  Cities        049       25113       30400   
56  224      Taylorsville  Cities        142       59989       60192   
57  225       Tooele City  Cities        048       32629       35251   
58  228         Tremonton  Cities        113        7828        8882   
59  232            Vernal  Cities        024        9460       10370   
60  238   Washington City  Cities        027       19640       27686   
61  243    West Bountiful  Cities        061        5420        5731   
62  245       West Jordan  Cities        155      105891      116046   
63  247  West Valley City  Cities        167      133796      136401   
64  252       Woods Cross  Cities        057       10086       11328   

    AREA_SQMI                                              SHAPE  \
55  10.193900  {'rings': [[[411406.60675760085, 4549994.30436...   
56  10.839700  {'rings': [[[422191.826826254, 4504303.3642982...   
57  24.087099  {'rings': [[[392501.67668449413, 4486111.93423...   
58   7.955100  {'rings': [[[399303.9767575987, 4621607.124490...   
59   4.632710  {'rings': [[[627188.8071560834, 4479658.624106...   
60  36.165600  {'rings': [[[278123.416362615, 4102773.6636791...   
61   3.299440  {'rings': [[[421904.10680578306, 4528452.80425...   
62  32.306999  {'rings': [[[415038.796736156, 4499131.1642154...   
63  35.799702  {'rings': [[[418708.9467166035, 4508796.864202...   
64   3.835550  {'rings': [[[424779.70679974474, 4524289.10421...   

    L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
55             35000.0             83233.0            1000.0   
56            350000.0            448861.0           15000.0   
57            963312.0            292869.0           30000.0   
58            908501.0            988499.0           15000.0   
59           1528421.0            133749.0            6000.0   
60           2405747.0            236874.0          600000.0   
61            500000.0             28261.0           30000.0   
62           1507571.0           1217193.0          350000.0   
63           6342247.0           3226747.0           40000.0   
64           1213221.0            514717.0            1000.0   

    L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  S_FINANCIAL_19  \
55          1006625.0           215900.0         17715.0        419955.0   
56          1363787.0          1374212.0        115555.0       2884476.0   
57          1076189.0          2013166.0         27658.0        901999.0   
58           450000.0          1055177.0          4000.0        149419.0   
59           200000.0            60406.0         11153.0        447033.0   
60           800000.0          1143016.0          5416.0        452707.0   
61           150000.0           176397.0          7000.0         82834.0   
62          2435088.0         10656831.0        146000.0       2515502.0   
63          4284059.0         27433099.0        303603.0       3288592.0   
64             1000.0           943585.0         81050.0        231516.0   

    L_RSTRNT_BAR_19  H_HEALTH_CARE_19  S_IT_DATA_19  S_CORPORATE_MGMT_19  \
55        4513352.0           24345.0     3145080.0              40000.0   
56       29151676.0          147429.0    10615060.0               9000.0   
57       12216749.0         1062603.0     4658818.0               4000.0   
58        2987225.0            6000.0      988901.0               2000.0   
59        9682438.0          346663.0     5079678.0               6000.0   
60        8988330.0            6000.0     3150837.0              35000.0   
61        3800517.0            6000.0     1881065.0              15000.0   
62       38940861.0         1026311.0    23262541.0             300000.0   
63       55761212.0         1545954.0    29270614.0              60000.0   
64        4982772.0           35000.0     2865960.0               2000.0   

    M_MANUFACTURING_19  E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
55        

In [18]:
# Confirm join count, Salt Lake County (Unincorporated) should be the only one not joined
l1 = list(q4_2020['City'].value_counts().index)
l2 = list(cities_sdf['NAME'].value_counts().index)
list(set(l1) - set(l2))

['Salt Lake County (Unincorporated)']

In [19]:
# Create file gdb and export sdf to feature class
outputs = '.\\Outputs'

gdb = os.path.join(outputs, "taxable_sales.gdb")
if not arcpy.Exists(gdb):
    arcpy.CreateFileGDB_management(outputs, "taxable_sales.gdb")

cities_sdf2.spatial.to_featureclass(location=os.path.join(outputs, "taxable_sales.gdb","taxable_sales_utah_q4"))

'C:\\LocalGIS\\GitProjects\\UT_Sales_Tax_Analysis\\Outputs\\taxable_sales.gdb\\taxable_sales_utah_q4'

### Analysis

In [21]:
cities_sdf2

FID              NAME  SOURCE SALESTAXID  POPLASTCEN  POPLASTEST  \
0     5     American Fork  Cities        002       27413       32519   
1    12       Beaver City  Cities        002        3137        3104   
2    15          Blanding  Cities        002        3514        3696   
3    19         Bountiful  Cities        004       43480       44098   
4    21           Brigham  Cities        017       18277       19404   
..  ...               ...     ...        ...         ...         ...   
60  238   Washington City  Cities        027       19640       27686   
61  243    West Bountiful  Cities        061        5420        5731   
62  245       West Jordan  Cities        155      105891      116046   
63  247  West Valley City  Cities        167      133796      136401   
64  252       Woods Cross  Cities        057       10086       11328   

    AREA_SQMI                                              SHAPE  \
0   11.256600  {"rings": [[[434244.49682960496, 4467294.37416...   
1    6.638160  {"rings": [[[355522.91657235345, 4239879.43388...   
2   13.205500  {"rings": [[[634403.6869837723, 4163230.893496...   
3   13.209000  {"rings": [[[426520.60679261835, 4528697.30425...   
4   24.910801  {"rings": [[[415000.02685977484, 4598737.23444...   
..        ...                                                ...   
60  36.165600  {"rings": [[[278123.416362615, 4102773.6636791...   
61   3.299440  {"rings": [[[421904.10680578306, 4528452.80425...   
62  32.306999  {"rings": [[[415038.796736156, 4499131.1642154...   
63  35.799702  {"rings": [[[418708.9467166035, 4508796.864202...   
64   3.835550  {"rings": [[[424779.70679974474, 4524289.10421...   

    L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
0             554928.0            698435.0          200000.0   
1            1056763.0             20808.0           15000.0   
2             761898.0             61901.0            1000.0   
3             150000.0           2291194.0            2000.0   
4             984729.0            222487.0            1000.0   
..                 ...                 ...               ...   
60           2405747.0            236874.0          600000.0   
61            500000.0             28261.0           30000.0   
62           1507571.0           1217193.0          350000.0   
63           6342247.0           3226747.0           40000.0   
64           1213221.0            514717.0            1000.0   

    L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  S_FINANCIAL_19  \
0           1565976.0          9116585.0         58968.0       4184777.0   
1              9000.0            70000.0         25000.0        515603.0   
2             70000.0           276229.0        150000.0         21221.0   
3           1098539.0          1894207.0         91173.0       1332378.0   
4            260399.0           625648.0         67697.0        455220.0   
..                ...                ...             ...             ...   
60           800000.0          1143016.0          5416.0        452707.0   
61           150000.0           176397.0          7000.0         82834.0   
62          2435088.0         10656831.0        146000.0       2515502.0   
63          4284059.0         27433099.0        303603.0       3288592.0   
64             1000.0           943585.0         81050.0        231516.0   

    L_RSTRNT_BAR_19  H_HEALTH_CARE_19  S_IT_DATA_19  S_CORPORATE_MGMT_19  \
0        24894878.0          553191.0    10533285.0               9000.0   
1         2206640.0           45000.0      520147.0               3000.0   
2          450000.0           60000.0      431669.0               4000.0   
3        12786832.0         1039103.0     5303124.0               9000.0   
4         6659356.0          308659.0     2858043.0              25000.0   
..              ...               ...           ...                  ...   
60        8988330.0            6000.0     3150837.0              35000.0   
61        3800517.0            6

In [26]:
# Get year specific column names
standard_cols = ['FID','NAME','SOURCE','SALESTAXID','POPLASTCEN','POPLASTEST','AREA_SQMI', 'SHAPE']
names_19 = standard_cols + [col for col in cities_sdf2.columns if '_19' in col]
names_20 = standard_cols + [col for col in cities_sdf2.columns if '_20' in col]
names_D20 = standard_cols + [col for col in cities_sdf2.columns if '_D20' in col]

# subset columns by time
cities_q4_19 = cities_sdf2[names_19]
cities_q4_20 = cities_sdf2[names_20]
cities_q4_d20 = cities_sdf2[names_D20]


In [27]:
# 10 cities with smallest difference of revenue
cities_q4_d20[['NAME', 'POPLASTCEN', 'AREA_SQMI', 'TOTAL_D20']].sort_values('TOTAL_D20', ascending=False).head(10)

NAME  POPLASTCEN  AREA_SQMI    TOTAL_D20
54         St George       76548  77.209900  151971179.0
62       West Jordan      105891  32.306999   83155206.0
20            Layton       69457  22.576200   79218781.0
34              Orem       92993  18.628000   70010105.0
33             Ogden       85183  27.493500   63608629.0
23             Logan       49612  18.485800   61104463.0
48      South Jordan       51057  22.263201   55642063.0
47  Saratoga Springs       18394  23.145201   55161851.0
0      American Fork       27413  11.256600   53236533.0
63  West Valley City      133796  35.799702   51992910.0

In [29]:
# 10 cities with largest difference of revenue
cities_q4_d20[['NAME', 'POPLASTCEN', 'AREA_SQMI', 'TOTAL_D20']].sort_values('TOTAL_D20', ascending=True).head(10)

NAME  POPLASTCEN   AREA_SQMI   TOTAL_D20
32  North Salt Lake       16930    8.512150 -59910734.0
45   Salt Lake City      190749  110.734001 -48983396.0
43        Roosevelt        6260    6.757070  -3518293.0
57      Tooele City       32629   24.087099   -356201.0
30           Murray       47311   12.293700   -306705.0
28       Monticello        2053    3.349260   1218909.0
2          Blanding        3514   13.205500   1575504.0
9             Delta        3486    8.713920   1591043.0
35        Park City        7709   20.341700   1612619.0
31            Nephi        5544    4.741290   1861731.0

#### Some industries of interest:
- RETAIL-GASOLINE STATIONS (447)
- RETAIL-FOOD & BEVERAGE STORES (445)
- REAL ESTATE, RENTAL & LEASING (53)
- ARTS, ENTERTAINMENT AND RECREATION (71)
- ACCOMMODATION (721)